compositor_pipeline/src/pipeline/input/whip.rs

```diff
@@ -0,0 +1,248 @@
+use rand::Rng;
+use std::{
+    sync::{Arc, Mutex},
+    thread,
+};
+use tokio::sync::mpsc;
+
+use depayloader::{Depayloader, DepayloaderNewError};
+use tracing::error;
+
+use std::sync::atomic::AtomicBool;
+
+use crate::{
+    pipeline::{
+        decoder::{self},
+        encoder,
+        rtp::BindToPortError,
+        types::EncodedChunk,
+        whip_whep::InputConnectionUtils,
+        PipelineCtx,
+    },
+    queue::PipelineEvent,
+};
+use compositor_render::InputId;
+use crossbeam_channel::Sender;
+use tracing::{debug, span, Level};
+
+use super::{AudioInputReceiver, Input, InputInitInfo, InputInitResult, VideoInputReceiver};
+
+pub mod depayloader;
+
+#[derive(Debug, thiserror::Error)]
+pub enum WhipReceiverError {
+    #[error("Error while setting socket options.")]
+    SocketOptions(#[source] std::io::Error),
+
+    #[error("Error while binding the socket.")]
+    SocketBind(#[source] std::io::Error),
+
+    #[error("Failed to register input. Port: {0} is already used or not available.")]
+    PortAlreadyInUse(u16),
+
+    #[error("Failed to register input. All ports in range {lower_bound} to {upper_bound} are already used or not available.")]
+    AllPortsAlreadyInUse { lower_bound: u16, upper_bound: u16 },
+
+    #[error(transparent)]
+    DepayloaderError(#[from] DepayloaderNewError),
+}
+
+#[derive(Debug, Clone)]
+pub struct WhipReceiverOptions {
+    pub stream: WhipStream,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct InputVideoStream {
+    pub options: decoder::VideoDecoderOptions,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct InputAudioStream {
+    pub options: decoder::OpusDecoderOptions,
+}
+
+pub struct OutputAudioStream {
+    pub options: encoder::EncoderOptions,
+    pub payload_type: u8,
+}
+
+#[derive(Debug, Clone)]
+pub struct WhipStream {
+    pub video: Option<InputVideoStream>,
+    pub audio: Option<InputAudioStream>,
+}
```

Comment:
To me, name containing `Stream` is a bit misleading, it suggest some kind of stream of packets or sth like this, whereas it seems to be just input encoders and payload type options. Also, aren't `WhipReceiverOptions` and `WhipStream` basically the same?